feat: `atlas pull` for XBlock translations | FC-0012

OmarIthawi opened the pull request:
Implements the following proposal:

 - https://github.com/openedx/edx-platform/pull/33166

## Known limitations
 - [x] Test i18n translations:
    - [x] JavaScript
    - [x] Python 
 - [x] This won't pull other plugin translations such as `edx-val` and any non-XBlock library. A follow up pull request will be created.
 - [x] Address `translation.to_locale(lang_code)` vs `translation.to_locale(translation.get_language())`
 - [x] Fix `get_statici18n_js_url`
 - [x] Refactor settings location and shared variables
 - [x] Consider a dedicated `plugin_translations` edX Platform module independent of XBlock module
 - [x] Fix failed tests
 - [x] Update the method and directory names to match the python implementation in the `0019-oep-58-atlas-translations-design.rst` document
 - [x] Upgrade XBlock to 1.9.0: 
   - [x] https://github.com/openedx/XBlock/pull/691
   - [x] https://github.com/openedx/edx-platform/pull/33948 (done manually via this PR)
 - [x] Ensure non--verbose silent the output is respected in `pull_xblock_translations`, otherwise remove the command

## Testing instructions

Install the Drag and Drop v2 fork which supports OEP-58 translations:

```
devstack $ make cms-shell
edx-platform # pip uninstall xblock-drag-and-drop-v2
edx-platform # pip install -e https://github.com/Zeit-Labs/xblock-drag-and-drop-v2/archive/atlas-translations.zip
edx-platform # make OPENEDX_ATLAS_PULL=yes pull_xblock_translations
devstack $ make cms-restart
```

Then add a Drag and Drop XBlock instance and it should use the translations from the `atlas pull` results instead of the bundled ones. 

Of course, it'll be difficult to distinguish which is which, therefore I've hacked the process as described in the `Testing results`.

## Testing results

I've manually replaced the `atlas pull` French translations with Chinese so I ensure that I'm not just seeing stale translations.

This is a temp. hack during testing and it looks like the Drag and Drop v2 translation is being pulled correctly via atlas.

These are the lines that I've added after the `atlas pull` command:

```py
def atlas_pull_xblock_by_module_names():
    ...
    # This hack purpose is to make it obvious that the translations are working by showing Chinese instead of Arabic.
    for dummy_lang in ['ar', 'fr']:
        subprocess.run(
            args=f'rm -rf {dummy_lang} && cp -r zh_CN {dummy_lang}',
            shell=True,
            cwd=f'{XBLOCK_TRANSLATIONS_ROOT}/drag_and_drop_v2/conf/locale',
            check=True,
        )
```

<details><summary>Screenshots</summary>

**Python code translations**
![Screenshot from 2023-11-12 22-26-02](https://github.com/openedx/edx-platform/assets/645156/cb5d6cd1-c2cd-4632-8199-a3066bfb201c)

**JavaScript code translations**
![Screenshot from 2023-11-12 22-26-20](https://github.com/openedx/edx-platform/assets/645156/2e0616dd-50b7-4795-b332-0e00d3691453)

</details> 

## Refs
This pull request is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which implements the [Translation Infrastructure update OEP-58](https://docs.openedx.org/en/latest/developers/concepts/oep58.html).

OmarIthawi committed:
feat: `atlas pull` for XBlock translations

openedx-webhooks commented:
Thanks for the pull request, @OmarIthawi! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

feanil commented on common/djangoapps/xblock_django/api.py:
```diff
@@ -2,6 +2,7 @@
 API methods related to xblock state.
 """
 
+import pkg_resources
```
`pkg_resources` is technically deprecated, can you use the new `importlib.resources` tooling instead?

https://setuptools.pypa.io/en/latest/pkg_resources.html

feanil commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,203 @@
+    xblock_domain_mo_file = locale_messages_dir / 'xblock.mo'
+    shutil.copyfile(source_mo_file, xblock_domain_mo_file)
+
+    # TODO: Consider using statici18n.management.commands.compilejsi18n
```
What's the plan with this todo? Is it something you plan on resolving before we merge this PR?

OmarIthawi replied:
I will give it a try. If it doesn't work, I'll remove the TODO.

feanil commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,203 @@
+
+def get_javascript_i18n_file(xblock_module, locale):
+    """
+    TBD
```
Note: need a real docstring here.

feanil commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,203 @@
+
+
+def get_javascript_i18n_file_path(xblock_module, locale):
+    return settings.STATICI18N_ROOT / get_javascript_i18n_file(xblock_module, locale)
```
Here you're returning a `path` type but above you're returning a string, is there a reason for that?

OmarIthawi replied:
Generally, `Path` and `str` are interchangeable and `get_javascript_i18n_file_name` is used for URLs as well as file system paths. Therefore, it makes sense to use `str` imo.

feanil commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,203 @@
+    )
+
+
+def get_javascript_i18n_file(xblock_module, locale):
```
Suggestion: Since you're not returing a `File` object, maybe this should be suffixed with `_name` to follow the same pattern as the function below that returns a `_path`
```suggestion
def get_javascript_i18n_file_name(xblock_module, locale):
```

feanil commented on xmodule/modulestore/django.py:
```diff
@@ -365,34 +367,104 @@ class XBlockI18nService:
     i18n service.
 
+    This service supports OEP-58 translations (https://docs.openedx.org/en/latest/developers/concepts/oep58.html)
+    that's pulled via atlas.
```
```suggestion
    that are pulled via atlas.
```

Grammer nit.

feanil commented on xmodule/modulestore/django.py:
```diff
@@ -365,34 +367,104 @@ class XBlockI18nService:
     """
     def __init__(self, block=None):
         """
         Attempt to load an XBlock-specific GNU gettext translator using the XBlock's own domain
```
```suggestion
        Attempt to load an XBlock-specific GNU gettext translations using the XBlock's own domain
```

feanil commented on xmodule/modulestore/django.py:
```diff
@@ -365,34 +367,104 @@ class XBlockI18nService:
-            xblock_locale_dir = 'translations'
-            xblock_locale_path = resource_filename(xblock_resource, xblock_locale_dir)
-            xblock_domain = 'text'
+            xblock_locale_domain, xblock_locale_dir = self.get_python_locale_dir(block)
```
Since this is now a class method you should call it from the class and not using an instance of the class.

feanil commented on xmodule/waffle.py:
```diff
@@ -0,0 +1,20 @@
+from edx_toggles.toggles import WaffleSwitch
+
+# XModule Namespace
+WAFFLE_NAMESPACE = 'xmodule'
```
I understand that we still call this part of the codebase the xmodule but given that we're skipping xmodules specifically, does it make more sense to namespaces this waffle flag to the `xblock` namespace?

OmarIthawi replied:
That'll be removed. I've designed the code to _not_ need a switch flag.

OmarIthawi commented on common/djangoapps/xblock_django/api.py:
```diff
@@ -27,6 +28,15 @@ def disabled_xblocks():
     return XBlockConfiguration.objects.current_set().filter(enabled=False)
 
 
+# TODO: move to openedx.core.xblock.api??
```
@kdmccormick we're adding new functions to the XBlock API module. Should we add it here, or in the new `openedx.core.xblock.api` v2 module?

OmarIthawi replied:
TODO: Consider using https://github.com/openedx/XBlock/blob/master/xblock/plugin.py instead.

OmarIthawi commented on common/djangoapps/xblock_django/management/commands/pull_plugin_translations.py:
```diff
@@ -0,0 +1,29 @@
+class Command(BaseCommand):
+
+    def add_arguments(self, parser):
+        parser.add_argument(
```
TODO: Ensure non-`--verbose` silent the output.

OmarIthawi commented on xmodule/modulestore/django.py:
```diff
@@ -365,34 +366,104 @@ class XBlockI18nService:
+        deprecated_xblock_locale_path = resource_filename(xblock_module_name, 'translations')
+        return 'text', deprecated_xblock_locale_path
+
+    @classmethod
```
Move the paths functions into a `xmodule/modulestore/api.py` module which can be imported by `common/xblock_django`.

OmarIthawi commented on common/djangoapps/xblock_django/management/commands/list_plugin_modules.py:
```diff
@@ -0,0 +1,12 @@
+
+class Command(BaseCommand):
+    def handle(self, *args, **options):
+        for xblock_module_name in get_xblock_module_names():
```
TODO: XBlock entry point name differs in Drag and Drop XBlock:

```
'xblock.v1': 'drag-and-drop-v2 = drag_and_drop_v2:DragAndDropBlock'
```

Which one should we use.

`drag_and_drop_v2` is the python module name.
`drag-and-drop-v2` is the `xblock.tag` variable and the entry name.

OmarIthawi commented on .gitignore:
```diff
@@ -61,6 +61,8 @@ conf/locale/fake*/LC_MESSAGES/*.po
 # this was a mistake in i18n_tools, now fixed.
 conf/locale/messages.mo
+conf/plugins-locale/
+/*/static/js/i18n/xblock.v1/
```
TODO: Use one of the following paths:

 - lms/static/js/i18n-xblock.v1/
 - lms/static/js/xblock.v1-i18n/

OmarIthawi commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,209 @@
+    return None
+
+
+def compile_xblock_messages():
```
TODO: Move this into the XBlock package.

OmarIthawi commented on common/djangoapps/xblock_django/plugin_translations.py:
```diff
@@ -0,0 +1,209 @@
+
+def create_js_namespaced_catalog(xblock_conf_locale_dir, locale, source_mo_file, namespace, output):
+    """
+    Compile .mo to .js gettext catalog similar to the `compilejsi18n` command.
```
TODO: Check the license and consider contributing this function into https://github.com/zyegfryed/django-statici18n instead of keeping it here.

OmarIthawi commented:
@feanil @kdmccormick Thanks a lot for the code review yesterday!

I've addressed most of the review comments already and added the rest as TODO check list items in the GitHub pull request description. Feel free to take a look now or later after tests are fixed.

I'll fix tests and tag you for a review before I start adding new test cases.

cc: @brian-smith-tcril

feanil reviewed (approved):
Changes look good to me, I like the more general management command names also so it will be good to update the docs to match as you are already planning to do.

OmarIthawi commented:
Thanks @feanil for the review. I'll finish the remainder TODOs, add tests and let you know when I'm done for a final review.

OmarIthawi committed:
chore: bump to xblock[django]==1.9.0

OmarIthawi commented:
@feanil this is ready again for review. I've added tests and refactored the work into reusable modules for future work (within FC-0012).

feanil commented:
Hey Omar,  sounds good.  FYI, I'm going on vacation for the next week so I probably won't get to this till the new year but I'll review it as soon as I can after I'm back.  Looks like there are still some test failures to deal with.  Let me know if you want me to wait for fixes or if I should start review as soon as I can.

OmarIthawi commented:
> Hey Omar, sounds good. FYI, I'm going on vacation for the next week so I probably won't get to this till the new year but I'll review it as soon as I can after I'm back. Looks like there are still some test failures to deal with. Let me know if you want me to wait for fixes or if I should start review as soon as I can.

Understood. If it's not inconvenient, please start the review while I'm fixing the test failures.

I think they're mostly easy test fixes.

OmarIthawi commented:
@feanil I've tested this and it seems to be working fine and I think it's ready for a final review.

I've thought about the `openedx/core/djangoapps/plugins/i18n_api.py` module which probably should live in the `openedx-atlas` package, but this package has no Python code and the feature is mostly Open edX specific since it's just passing parameters around, so I kept it in the `edx-platform`.

feanil commented on Makefile:
```diff
@@ -55,8 +56,19 @@ endif
 	i18n_tool transifex push
 
-pull_translations:  ## pull translations from Transifex
+pull_xblock_translations:
```
```suggestion
pull_xblock_translations: ## pull xblock translations via atlas
```

This comment becomes the help text for `make help` so it's useful to have something.

feanil commented on openedx/core/djangoapps/plugins/tests/test_i18n_api.py:
Is this file supposed to be empty?

OmarIthawi replied:
no, i forgot to add test for this one

feanil reviewed (commented):
Generally looks like it's ready to go, I had one question about an empty test file.

feanil commented:
@OmarIthawi it looks like this is good to go, is there anything else we need to do before we merge this?

OmarIthawi commented:
@feanil I think it's ready, I've pushed it with few more tests in `test_i18n_api.py`.

OmarIthawi commented:
@feanil @brian-smith-tcril A gentle reminder that this pull request is ready to be merged.

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -181,6 +167,77 @@ Introduce new Django commands to the ``edx-platform``:
+
+#. **The XBlock/plugin tag:** Python plugins has an entry point name in which
+   is referred to as ``tag`` in Open edX. For example, the
+   the tag the `Drag and Drop XBlock setup.py file`_ is ``drag-and-drop-v2``::
```
```suggestion
#. **The XBlock/plugin tag:** Python plugins have an entry point name which
   is referred to as ``tag`` in Open edX. For example, the
   tag in the `Drag and Drop XBlock setup.py file`_ is ``drag-and-drop-v2``::
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -181,6 +167,77 @@ Introduce new Django commands to the ``edx-platform``:
+instead of the repository name to pull translations from the
+`openedx-translations repo`_ via ``atlas``.
+
+This trade-off has the following pros and cons:
```
```suggestion
Using the Python module name has the following pros and cons:
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -181,6 +167,77 @@ Introduce new Django commands to the ``edx-platform``:
+  the python module name another unique identifier for XBlocks.
+
+The trade-offs are acceptable and this decision is reversible in case
+the ``xblock.tag`` needed to be used. However, this will require parsing
```
```suggestion
the ``xblock.tag`` needs to be used. However, this will require parsing
```

brian-smith-tcril reviewed (approved):
Left a couple small docs suggestions but overall this LGTM!

feanil commented:
@OmarIthawi if the suggestions look good to you, can you add them in?  I can merge this PR tomorrow morning our time.

feanil commented:
Looks like this needs a rebase and conflict resolution as well now that the other atlas PR landed.

OmarIthawi commented on Makefile:
```diff
@@ -65,13 +74,13 @@ ifeq ($(OPENEDX_ATLAS_PULL),)
+	make pull_xblock_translations
 	find conf/locale -mindepth 1 -maxdepth 1 -type d -exec rm -r {} \;
-	atlas pull $(OPENEDX_ATLAS_ARGS) translations/edx-platform/conf/locale:conf/locale
+	atlas pull $(ATLAS_OPTIONS) translations/edx-platform/conf/locale:conf/locale
```
Thanks @feanil for the quick triage.

Conflicts has been resolved and I've tested it again and it seems to be playing well. 

I've changed this line to match the Tutor option to have one less surprise for operators.

openedx-webhooks commented:
@OmarIthawi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/33698&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2023-11-10+23:05&entry.1917517419=2024-01-12+14:50&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
